src/libfuncs/function_call.rs

```diff
@@ -161,26 +137,62 @@ pub fn build<'ctx, 'this>(
                 let val = cont_block.argument(count)?.into();
                 count += 1;
 
-                results.push(if type_info.is_memory_allocated(registry) {
-                    let ty = type_info.build(context, helper, registry, metadata, &var_info.ty)?;
-                    let layout = type_info.layout(registry)?;
-
-                    let stack_ptr =
-                        helper
-                            .init_block()
-                            .alloca1(context, location, ty, Some(layout.align()))?;
-
-                    cont_block.store(context, location, stack_ptr, val, Some(layout.align()));
-
-                    stack_ptr
-                } else {
-                    val
-                });
+                results.push(val);
             }
         }
 
         cont_block.append_operation(helper.br(0, &results, location));
     } else {
+        let mut result_types = Vec::new();
+        let return_types = info
+            .function
+            .signature
+            .ret_types
+            .iter()
+            .filter_map(|type_id| {
+                let type_info = registry.get_type(type_id).unwrap();
+                if type_info.is_builtin() && type_info.is_zst(registry) {
+                    None
+                } else {
+                    Some((type_id, type_info))
+                }
+            })
+            .collect::<Vec<_>>();
+        let has_return_ptr = if return_types.len() > 1 {
```

Comment:
can you add a high-level description of each branch here? It can improve readability since the condition on each branch are kind of large.